job_definitions/release_application_paas.yml

```diff
@@ -53,14 +53,23 @@
               }
 
               stage('Deploy') {
-                withEnv(["DM_CREDENTIALS_REPO=/var/lib/jenkins/digitalmarketplace-credentials", "CF_HOME=${pwd()}"]) {
-                  paas_credentials = sh(script: '$DM_CREDENTIALS_REPO/sops-wrapper -d $DM_CREDENTIALS_REPO/jenkins-vars/paas_credentials_env.enc', returnStdout: true).trim()
-                  withEnv(paas_credentials.tokenize("\n")) {
-                      sh('make paas-login')
+                  withEnv(["DM_CREDENTIALS_REPO=/var/lib/jenkins/digitalmarketplace-credentials", "CF_HOME=${pwd()}"]) {
+                    paas_credentials = sh(script: '$DM_CREDENTIALS_REPO/sops-wrapper -d $DM_CREDENTIALS_REPO/jenkins-vars/paas_credentials_env.enc', returnStdout: true).trim()
+                    withEnv(paas_credentials.tokenize("\n")) {
+                        sh('make paas-login')
+                    }
                   }
-                  lock("paas-${STAGE}-${APPLICATION_NAME}-deploy") {
-                      sh('make ${STAGE} deploy-app')
+
+                lock("paas-${STAGE}-${APPLICATION_NAME}-deploy") {
+                  if ("$RELEASE_NAME" == "") {
+                    echo "No release specified. Detecting currently deployed release version of ${APPLICATION_NAME} ..."
+
+                    RELEASE_NAME = sh(script: "curl -s https://\$(cf curl /v2/apps/\$(cf app ${APPLICATION_NAME} --guid)/routes | jq -crM '.resources[0].entity.host').cloudapps.digital\$(cf curl /v2/apps/\$(cf app ${APPLICATION_NAME} --guid)/routes | jq -crM '.resources[0].entity.path')/_status | jq -crM '.version'", returnStdout: true)
```

Comment:
This line has a lot going on! Let me check if I understand the different parts of what's happening here:
- Get the Cloud Foundry app ID with `cf app ${APPLICATION_NAME} --guid` (twice?)
- Use that app ID to get the resource host and path, via two separate `curl` requests
- Construct the url to get that resource's `/_status`, to get the latest `version`

I think splitting out fetching the app ID into a separate var might make this more readable (that's assuming I've read all this right!)

Reply:
Added a comment to hopefully clear things up a little.

Painful :(